### ISSUE TITLE
vips_colourspace: no known route from 'cmyk' to 'srgb' 

### ISSUE THREAD

**Issue by mmhan** [janko/image_processing#79]
I am using this gem together with rails' ActiveStorage and facing an error with the message `vips_colourspace: no known route from 'cmyk' to 'srgb'`. 

I'm not sure how I can resolve that issue.

**Comment by mikelieser** [comment]
Same here. A user did upload a CMYK jpg (which is totally fine in that use case, as it is intended to be downloaded by others, not to display on web) and according to ActiveStorage content type definitions **image/jpeg** is (of course) representable, but it fails producing the preview in vips with...

`vips_colourspace: no known route from 'cmyk' to 'srgb'`

On MacOS 11.4 with image_processing (1.12.1) and vips (8.10.6) via brew it does work with all test files. But on the server with ubuntu docker and libvips the errors described happening. But not for all uploaded CMYK images. The ones I exported in Photoshop are working.

Ruby: 3.0.1

**Comment by mikelieser** [comment]
@janko Do you think this is something that needs to be handled in **image_processing** or https://github.com/libvips/ruby-vips ?

https://libvips.github.io/libvips/API/8.7/Using-vipsthumbnail.md.html #colour management

**Comment by Flixt** [comment]
The issue is already fixed in libvips. Upgrading libvips to > `8.10` on the server would solve it for us. Sadly we run Debian Buster there and the latest version for libvips that is available is `8.7.x`.

It is possible to compile libvips by hand on the server and install it, but this interferes with the installed package
and it can be tedious to make sure that all dependencies are available for all file formats. So we did the following.

1. Place a `cmyk.icc` profile somewhere on the server, this could be one you already have, e.g. when you install `ghostscript` on the server there is one at `/usr/share/color/icc/ghostscript/default_cmyk.icc` (Debian). Or you could take the one from https://github.com/libvips/libvips/tree/master/libvips/colour/profiles (you should check the license in that case though).

2. Add the following monkey patch and adapt the path to the profile from step 1

`config/initializers/vips_fix.rb`
```ruby
module ImageProcessing
  module Vips
    extend Chainable

    class Processor < ImageProcessing::Processor
      private

      alias_method :original_thumbnail, :thumbnail
      def thumbnail(width, height, sharpen: SHARPEN_MASK, **options)
        original_thumbnail(width, height, sharpen: sharpen, **options)
      rescue ::Vips::Error => e
        if e.message.include?("no known route from 'cmyk' to 'srgb'")
          options[:import_profile] = Rails.root.join('lib', 'colors', 'cmyk.icc').to_s # adapt the path here
          original_thumbnail(width, height, sharpen: sharpen, **options)
        else
          raise e
        end
      end
    end
  end
end
```

Seems hacky, but it solves the problem for us until we can update to a newer version of libvips.

**Comment by mikelieser** [comment]
@Flixt Thanks! The monkey patch is working great. The latest version that is available for us is `8.9.x` :)

**Comment by janko** [comment]
Given that this was a libvips issue that has since been fixed, I'm closing this one.
